### PR TITLE
[kernel] MiniMax-Text-01 decode lightning_attn with triton

### PR DESCRIPTION
**Pull request by BBuf**
Currently, a simple Triton kernel has been implemented to perform kernel fusion for MiniMax-Text-01  Lightning Attention decoding. However, this version **is not efficient** , possibly due to the extra computations after padding and the padding process itself. Nonetheless, it can serve as a baseline version for precision alignment and optimization. Subsequent optimizations can be based on this benchmark script.

h200 result:

```shell
lightning-attention-decode-performance:
    batch_size  seq_len  Original PyTorch Implementation  Triton Implementation
0          1.0      1.0                       378.304005             313.455999
1          2.0      1.0                       373.984009             311.551988
2          4.0      1.0                       375.104010             312.480003
3          8.0      1.0                       373.216003             309.664011
4         16.0      1.0                       369.311988             339.071989
5         32.0      1.0                       459.183991             461.616009
6         64.0      1.0                       720.640004             713.248014
7        128.0      1.0                      1254.863977            1230.479956
8        256.0      1.0                      2341.664076            2286.751986
9        512.0      1.0                      4521.664143            4410.896301
10      1024.0      1.0                      8960.543633            8729.727745
11      2048.0      1.0                     17803.615570           17333.471298
```

### CODE REVIEW

**Comment by DragonFive** [comment]
Can lightning_attn_decode run in a tensor_parallel manner by placing different heads on different ranks?

**Comment by BBuf** [comment]
> Can lightning_attn_decode run in a tensor_parallel manner by placing different heads on different ranks?

The matmul operations involved here are actually gemv (General Matrix-Vector Multiplication) rather than gemm (General Matrix-Matrix Multiplication), so I believe they are unnecessary.

```python
def inference(
        self,
        x,
        attn_mask: Optional[torch.Tensor] = None,  # (b, n)
        output_attentions: bool = False,
        past_key_value: Optional[Tuple[torch.Tensor]] = None,
        use_cache: bool = False,
        slope_rate: Optional[torch.Tensor] = None,  # (h, 1, 1)
    ):
        # x: b n d
        b, n, d = x.shape
        # linear map
        qkv = self.act(self.qkv_proj(x))
        new_shape = qkv.size()[:-1] + (self.num_heads, -1)
        qkv = qkv.view(*new_shape)
        q, k, v = torch.split(qkv, [self.head_dim] * 3, dim=3)
        q = q.transpose(1, 2) # [b, n, h, d] -> [b, h, n, d]
        k = k.transpose(1, 2) # [b, n, h, d] -> [b, h, n, d]
        v = v.transpose(1, 2) # [b, n, h, d] -> [b, h, n, d]

        self.offset += 1
        ratio = torch.exp(-slope_rate) #[h, 1, 1]

        # decode mode
        kv = past_key_value #[b, h, d, e]
        output = []
        for i in range(n):
            # kv: [b, h, d, e]
            # ratio: [h, 1, 1]
            # k: [b, h, n, d]
            # v: [b, h, n, e]
            # k[:, :, i : i + 1]: [b, h, 1, d]
            # v[:, :, i : i + 1]: [b, h, 1, e]
            # ratio * kv: [b, h, d, e]
            # torch.einsum(
            #     "... n d, ... n e -> ... d e",
            #     k[:, :, i : i + 1],
            #     v[:, :, i : i + 1],
            # )
            # [b, h, d, e] + [b, h, d, e] -> [b, h, d, e]
            kv = ratio * kv + torch.einsum(
                "... n d, ... n e -> ... d e",
                k[:, :, i : i + 1],
                v[:, :, i : i + 1],
            )
            # q[:, :, i : i + 1]: [b, h, 1, d]
            # kv.to(q.dtype): [b, h, d, e]
            # torch.einsum(
            #     "... n e, ... e d -> ... n d", q[:, :, i : i + 1], kv.to(q.dtype)
            # )
            # [b, h, 1, e] * [b, h, d, e] -> [b, h, 1, d]
            qkv = torch.einsum(
                "... n e, ... e d -> ... n d", q[:, :, i : i + 1], kv.to(q.dtype)
            )
            output.append(qkv)
        output = torch.concat(output, dim=-2)

        # reshape
        output = rearrange(output, "b h n d -> b n (h d)")
        # normalize
        output = self.norm(output)
        # gate
        output = F.sigmoid(self.output_gate(x)) * output
        # outproj
        output = self.out_proj(output)

        attn_weights = None

        return output, attn_weights, kv
```

**Review by merrymercy — APPROVED** [review]
